daemons/execd/remoted_schemas.c

```diff
@@ -0,0 +1,229 @@
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <crm/cib.h>
+#include <crm/cib/cib_types.h>
+#include <crm/cib/internal.h>
+#include <crm/crm.h>
+#include <crm/msg_xml.h>
+#include <crm/common/mainloop.h>
+
+#include "pacemaker-execd.h"
+
+static int
+rm_files(const char *pathname, const struct stat *sbuf, int type, struct FTW *ftwb)
+{
+    /* Don't delete PCMK__REMOTE_SCHEMA_DIR . */
+    if (pcmk__str_eq(pathname, PCMK__REMOTE_SCHEMA_DIR, pcmk__str_none)) {
+        return 0;
+    }
+
+    if (remove(pathname) != 0) {
+        int rc = errno;
+        crm_err("Could not remove %s: %s", pathname, pcmk_rc_str(rc));
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+clean_up_extra_schema_files(void)
+{
+    struct stat sb;
+    int rc;
+
+    rc = stat(PCMK__REMOTE_SCHEMA_DIR, &sb);
+
+    if (rc == -1) {
+        if (errno == ENOENT) {
+            /* If the directory doesn't exist, try to make it first. */
+            if (mkdir(PCMK__REMOTE_SCHEMA_DIR, 0755) != 0) {
+                rc = errno;
+                crm_err("Could not create directory for schemas: %s",
+                        pcmk_rc_str(rc));
+            }
+
+        } else {
+            rc = errno;
+            crm_err("Could not create directory for schemas: %s",
+                    pcmk_rc_str(rc));
+        }
+
+    } else if (!S_ISDIR(sb.st_mode)) {
+        /* If something exists with the same name that's not a directory, that's
+         * an error.
+         */
+        crm_err("%s already exists but is not a directory", PCMK__REMOTE_SCHEMA_DIR);
+
+    } else {
+        /* It's a directory - clear it out so we can download potentially new
+         * schema files.
+         */
+        rc = nftw(PCMK__REMOTE_SCHEMA_DIR, rm_files, 10, FTW_DEPTH|FTW_MOUNT|FTW_PHYS);
+
+        if (rc != 0) {
+            crm_err("Could not remove %s: %s", PCMK__REMOTE_SCHEMA_DIR, pcmk_rc_str(rc));
+        }
+    }
+}
+
+static void
+write_extra_schema_file(xmlNode *xml, void *user_data)
+{
+    const char *file = NULL;
+    char *path = NULL;
+    int rc;
+
+    file = crm_element_value(xml, PCMK__XA_PATH);
+    if (file == NULL) {
+        crm_warn("No destination path given in schema request");
+        return;
+    }
+
+    path = crm_strdup_printf("%s/%s", PCMK__REMOTE_SCHEMA_DIR, file);
+
+    /* The schema is a CDATA node, which is a child of the <file> node.  Traverse
+     * all children and look for the first CDATA child.  There can't be more than
+     * one because we only have one file attribute on the parent.
+     */
+    for (xmlNode *child = xml->children; child != NULL; child = child->next) {
+        FILE *stream = NULL;
+
+        if (child->type != XML_CDATA_SECTION_NODE) {
+            continue;
+        }
+
+        stream = fopen(path, "w+");
+        if (stream == NULL) {
+            crm_warn("Could not write schema file %s: %s", path, strerror(errno));
+        } else {
+            rc = fprintf(stream, "%s", child->content);
```

Comment:
When you test this, do an md5sum on the downloaded schemas to see if they're identical. It's possible there may be whitespace differences or whatnot but that's worth checking.

Reply:
Are you looking for comparing the md5sum of child->content against the md5sum that the server said the content should be?  If that's the case, I'll need to add it as an attribute to the <file> node.  I don't think we add that automatically anywhere at the moment.

Reply:
No, not in the code, just as a manual comparison for testing purposes

Reply:
These all look good.